core/lib/resources/presence/sentry.js

```diff
@@ -1,159 +1,270 @@
+  try { 
+    return JSON.parse(message); 
   } catch (e) { }
-}
+};
+
+var messageIsExpired = function(message) {
+  return (!message || !message.expiration || (message.expiration <= Date.now()));
+};
 
-function Sentry(name) {
-  this.name = name;
+var messageExpiration = function(message) {
+  return (message && message.expiration && (message.expiration - Date.now()));
+};
+
+var Sentry = function(name, options) {
   this.sentries = {};
-}
+  this._setName(name);
 
-Sentry.expiry = 20000;
-Sentry.channel = 'sentry:/radar';
+  this._defaultExpiryOffset = defaultOptions.DEFAULT_EXPIRY_OFFSET;
+  this._refreshInterval = defaultOptions.REFRESH_INTERVAL;
+  this._checkInterval = defaultOptions.CHECK_INTERVAL;
+
+  this._applyOptions(options);
+};
 
 require('util').inherits(Sentry, require('events').EventEmitter);
 
-Sentry.prototype._listen = function() {
-  var self = this;
-  if (!this.listener) {
-    this.listener = function(channel, m) {
-      if (channel != Sentry.channel) {
-        return;
-      }
-      self._save(parse(m));
-    };
-    Persistence.pubsub().on('message', this.listener);
+Sentry.prototype.start = function(options, callback) {
+  var self = this,
+      keepAliveOptions = {},
+      upMessage;
+  
+  options = options || {};
+
+  if (typeof(options) === 'function') { 
+    callback = options; 
+  } else {
+    this._applyOptions(options); 
   }
-  if (!this.subscribed) {
-    this.subscribed = true;
-    Persistence.pubsub().subscribe(Sentry.channel);
+
+  if (this._refreshTimer) { return; }
+
+  logging.info('#presence - #sentry - starting', this.name);
+
+  if (options.expiration) {
+    keepAliveOptions.expiration = options.expiration;
   }
+
+  upMessage = self._keepAlive(keepAliveOptions);
+
+  this.setMaxListeners(0);
+  this._loadAndCleanUpSentries(function() {
+    self.emit('up', this.name, upMessage);
+    Persistence.publish(redisSentriesKey, upMessage);
+    self._startListening();
+    if (callback) { callback(); }
+  });
+
+  this._refreshTimer = setTimeout(this._refresh.bind(this), Math.floor(this._refreshInterval/2));
+  this._checkSentriesTimer = setTimeout(this._checkSentries.bind(this), Math.floor(this._checkInterval/2));
 };
 
+Sentry.prototype.stop = function(callback) {
+  logging.info('#presence- #sentry stopping', this.name);
+  
+  this._stopTimer('_checkSentriesTimer');
+  this._stopTimer('_refreshTimer');
+  this.sentries = {};
+  this._stopListening();
+
+  if (callback) { callback(); }
+};
 
-Sentry.prototype._save = function(message) {
-  if (message && message.name) {
-    logging.debug('#presence - sentry.save', message.name,
-                  (message.expiration - Date.now()) + '/' + Sentry.expiry);
-    this.sentries[message.name] = message;
+Sentry.prototype.sentryNames = function() {
+  return Object.keys(this.sentries);
+};
+
+Sentry.prototype.isDown = function(name) {
+  var lastMessage = this.sentries[name],
+      isSentryDown = messageIsExpired(lastMessage);
```

Comment:
I think redis saves us here from a race condition. I was thinking about a possibility where the sentry up has not been received, but a presence message with that sentry has arrived. But that's not possible due to ordering in redis messages.